components/constellation/constellation.rs

```diff
@@ -2474,53 +2384,111 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
+                    let session_history = self.joint_session_histories
+                        .entry(change.top_level_browsing_context_id).or_insert(SessionHistory::new());
+                    session_history.replace(replace_pipeline_id.clone(),
+                        AliveOrDeadPipeline::Alive(change.new_pipeline_id));
+
+                    match replace_pipeline_id {
+                        AliveOrDeadPipeline::Alive(pipeline_id) => vec![pipeline_id],
+                        AliveOrDeadPipeline::Dead(..) => vec![],
+                    }
+                } else {
+                    let session_history = self.joint_session_histories
+                        .entry(change.top_level_browsing_context_id).or_insert(SessionHistory::new());
+                    let diff = SessionHistoryDiff::BrowsingContextDiff {
+                        browsing_context_id: change.browsing_context_id,
+                        new_pipeline_id: AliveOrDeadPipeline::Alive(change.new_pipeline_id),
+                        old_pipeline_id: AliveOrDeadPipeline::Alive(old_pipeline_id),
+                    };
+
+                    session_history.push_diff(diff).into_iter()
+                        .map(|SessionHistoryDiff::BrowsingContextDiff { new_pipeline_id, .. }| new_pipeline_id)
+                        .filter_map(|pipeline_id| pipeline_id.alive_pipeline_id())
+                        .collect::<Vec<_>>()
+                };
+
+                for pipeline_id in pipelines_to_close {
+                    self.close_pipeline(pipeline_id, DiscardBrowsingContext::No, ExitPipelineMode::Normal);
+                }
+
+                self.update_activity(old_pipeline_id);
+                self.update_activity(change.new_pipeline_id);
+                self.notify_history_changed(change.top_level_browsing_context_id);
+            }
+        }
+
+        if let Some(top_level_id) = top_level_id {
+            self.trim_history(top_level_id);
         }
 
-        if new_context {
-            self.new_browsing_context(change.browsing_context_id,
-                                      change.top_level_browsing_context_id,
-                                      change.new_pipeline_id,
-                                      change.load_data);
-            self.update_activity(change.new_pipeline_id);
-            self.notify_history_changed(change.top_level_browsing_context_id);
+        self.update_frame_tree_if_active(change.top_level_browsing_context_id);
+    }
+
+    fn trim_history(&mut self, top_level_browsing_context_id: TopLevelBrowsingContextId) {
+        let pipelines_to_evict = {
+            let session_history = self.joint_session_histories.entry(top_level_browsing_context_id)
+                .or_insert(SessionHistory::new());
+
+            let history_length = PREFS.get("session-history.max-length").as_u64().unwrap_or(20) as usize;
+
+            let mut pipelines_to_evict = session_history.past.iter().rev()
```

Comment:
Add a comment about the use of `rev()` here (as in, I can't remember which order we get the session history past in, so this would help future me).

Reply:
Done.